Proposal.md

```diff
@@ -0,0 +1,604 @@
+
+```swift
+var array = ["a", "b", "c", "d", "e", "f", "g"]
+let newPosition = array.shift(from: 1, toJustBefore: 5)
+// array == ["a", "c", "d", "e", "b", "f", "g"]
+// newPosition == 4
+
+//    ┌─────┬─────┬─────┬─────┬─────┐┌─────┬─────┐
+//    │  a  │  b  │  c  │  d  │  e  ││  f  │  g  │
+//    └─────┴─────┴─────┴─────┴─────┘└─────┴─────┘
+//             ^                     ^
+//           source              insertion
+//                                 point
+//
+//    ┌─────┬─────┬─────┬─────┐┌─────┐┌─────┬─────┐
+//    │  a  │  c  │  d  │  e  ││  b  ││  f  │  g  │
+//    └─────┴─────┴─────┴─────┘└─────┘└─────┴─────┘
+//                                ^
+//                        newPosition == 4
+```
+
+The new gathering and shifting methods are listed below.
+
+```swift
+extension MutableCollection {
+    /// Collects the elements at the given indices just before the specified
+    /// index.
+    ///
+    /// This example finds all the uppercase letters in the array and gathers
+    /// them between `"i"` and `"j"`.
+    ///
+    ///     var letters = Array("ABCdeFGhijkLMNOp")
+    ///     let uppercase = letters.indices(where: { $0.isUppercase })
+    ///     let rangeOfUppercase = letters.gather(uppercase, justBefore: 10)
+    ///     // String(letters) == "dehiABCFGLMNOjkp"
+    ///     // rangeOfUppercase == 4..<13
+    ///
+    /// - Parameters:
+    ///   - indices: The indices of the elements to move.
+    ///   - insertionPoint: The index to use as the destination of the elements.
+    /// - Returns: The new bounds of the moved elements.
+    ///
+    /// - Complexity: O(*n* log *n*) where *n* is the length of the collection.
+    @discardableResult
+    public mutating func gather(
+        _ indices: RangeSet<Index>, justBefore insertionPoint: Index
+    ) -> Range<Index>
+
+    /// Collects the elements that satisfy the given predicate just before the
+    /// specified index.
+    ///
+    /// This example gathers all the uppercase letters in the array between
+    /// `"i"` and `"j"`.
+    ///
+    ///     var letters = Array("ABCdeFGhijkLMNOp")
+    ///     let rangeOfUppercase = letters.gather(justBefore: 10) { $0.isUppercase }
+    ///     // String(letters) == "dehiABCFGLMNOjkp"
+    ///     // rangeOfUppercase == 4..<13
+    ///
+    /// - Parameters:
+    ///   - predicate: A closure that returns `true` for elements that should
+    ///     move to `destination`.
+    ///   - insertionPoint: The index to use as the destination of the elements.
+    /// - Returns: The new bounds of the moved elements.
+    ///
+    /// - Complexity: O(*n* log *n*) where *n* is the length of the collection.
+    @discardableResult
+    public mutating func gather(
+        justBefore: Index, where predicate: (Element) -> Bool
```

Comment:
`justBefore` should have a parameter name that describes its role.  The parameters clause suggests `insertionPoint` which isn't bad.